File: Problems/Algorithms/209. Minimum Size Subarray Sum/MinSizeSubarray.kt
class Solution {
    fun minSubArrayLen(target: Int, nums: IntArray): Int {
        val n = nums.size
        var ans = n+1
        
        var sum = 0
        var j = 0
        for (i in 0..n-1) {
            sum += nums[i]
            if (sum >= target) {
                while (sum >= target && j <= i) {
                    sum -= nums[j]
                    j++
                }
                ans = minOf(ans, i-j+1)
            }
        }
        
        if (ans == n+1) return 0
        return ans+1
    }
}

File: Problems/Algorithms/209. Minimum Size Subarray Sum/min-size-subarray.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int left = 0, right = 0;
        int curr = 0;
        int ans = INT_MAX;

        for (right = 0; right < nums.size(); right++) {
            curr += nums[right];

            while (curr >= target) {
                ans = min(ans, right - left + 1);
                curr -= nums[left];
                left++;
            }
        }

        return ans == INT_MAX ? 0 : ans;
    }
};
